### PR TITLE
Bug fix 794 Fix of loop index, revidex and length issues

### PR DESCRIPTION
**Pull request by asenov**
This pull request fixes an issue https://github.com/pallets/jinja/issues/794, https://github.com/pallets/jinja/issues/244, https://github.com/pallets/jinja/issues/751 and issues related to revindex and length of iterable objects

### CODE REVIEW

File: jinja2/runtime.py [hunk 1]
```diff
@@ -422,29 +422,17 @@ class LoopContext(LoopContextBase):
...
+            # Getting length of iterable
+            iterable = tuple(iterable)
             self._length = len(iterable)
+            self._iterator = iter(iterable)
         except (TypeError, AttributeError):
             self._length = None
         self._after = self._safe_next()
 
     @property
     def length(self):
```
Comment (asenov) [inline jinja2/runtime.py]: Maybe this one can be moved to the base class? Since there in the parent class dependencies of this property?

**Comment by asenov** [comment]
So, any comments?

**Comment by asenov** [comment]
I suppose I have to fix failed tests?

**Comment by davidism** [comment]
Thanks for working on this! Closing in favor of #993.

**Pull request closed without merging**